text/0000-deployment-language.md

```diff
@@ -0,0 +1,561 @@
+It contains:
+
+ * an id that is unique per
+ * intrinsic connector type this artefact is based on
+ * generic connector config
+ * specific config for the connector type
+ * possibly a list of arguments that can/need to be provided upon creating/instantiating this artefact
+
+There are many artefacts like connectors, sources, sinks, pipelines, deployments.
+Here we want to describe 3 special artefacts: Connectors, Pipelines, Deployments:
+
+
+### Connectors
+
+EBNF Grammar:
+
+```
+connector       = "define" connector_type "connector" artefact_id
+                  [ "args" argument_list ]
+                  ( with_block | script_block )
+                  "end"
+connector_type  = [ module_path "::" ] id      
+module_path     = id [ "::" module_path ]
+artefact_id     = id     
+argument_list   = ( id | id "=" expr ) [ "," argument_list ]
+with_block      = "with" assignment_list
+assignment_list = assignment [ "," assignment_list ]
+assignment      = let_path "=" expr
+let_path        = id [ "." id ]
+script_block    = "script" tremor_script
+```
+
+Whitespace and newlines are not significant here.
+
+**Examples:**
+
+
+
+Every connector can be configured with a script block that returns a record value. This gives maximal flexibility.
+
+```
+define file connector my_file
+script
+    let interceptors = [ "newline" ],
+        codec = "string",
+        config.path = "/etc/passwd"
+end;
+```
+
+As it may be very verbose to define an artefact with a complicated script, we created the  `with` statement as a syntax sugar that creates a tremor-script block with a multi-let statement, which in turn returns a record value with all the defined variables.
+
+It is also possible to add `args` which can be provided upon artefact creation/instantiation. 
+
+```
+define http connector my_http
+args
+    required_argument,
+    optional_arg_with_default = "default value"
+with
+    codec = "json",
+    interceptors = [ 
+        {
+            "type": "split",
+            "config": {
+                "split_by": "\n"
+            }
+        },
+        "base64"
+    ],
+    config.host = "localhost",
+    config.port = args.required_argument,
+    err_required = args.optional_arg_with_default
+end;
+```
+
+
+#### Required Tremor Script Changes
+
+In order to make defining config entries in tremor-script convenient, we need to add a few features to tremor-script to make this work reasonably well in a configuration context, where all we want is to return a record value without too much fuss and ceremony:
+
+* Add multi-let statements, that combine multiple `let`s inside a single statement whcih defined the variables therein and returns a record value with all those definitions.
+* Add auto-creation of intermediary path segments in let stetements:
+
+  ```
+  let config.nested.value = 1;
+  ```
+
+  In this case if `config.nested` does not exist we would auto-create it as part of this let as an empty record.
```

Comment:
worth mentioning that it also that we error if we try to 'nest' into a existing non record field.